itests/gateway_test.go

```diff
@@ -297,3 +364,211 @@ func sendFunds(ctx context.Context, fromNode *kit.TestFullNode, fromAddr address
 
 	return nil
 }
+
+func TestGatewayRateLimits(t *testing.T) {
+	req := require.New(t)
+
+	kit.QuietMiningLogs()
+	ctx := context.Background()
+	tokensPerSecond := 10
+	requestsPerMinute := 30 // http requests
+	nodes := startNodes(ctx, t,
+		withNodeOpts(kit.DisableEthRPC()),
+		withPerConnectionAPIRateLimit(tokensPerSecond),
+		withPerHostRequestsPerMinute(requestsPerMinute),
+	)
+
+	nodes.full.WaitTillChain(ctx, kit.HeightAtLeast(10)) // let's get going first
+
+	// ChainHead uses chainRateLimitTokens=2.
+	// But we're also competing with the paymentChannelSettler which listens to the chain uses
+	// ChainGetBlockMessages on each change, which also uses chainRateLimitTokens=2. But because we're
+	// rate limiting, it only gets a ChainGetBlockMessages in between our ChainHead calls, not on each
+	// 5ms block (which it wants). So each loop should be 4 tokens.
+	loops := 25
+	tokensPerLoop := 4
+	start := time.Now()
+	for i := 0; i < loops; i++ {
+		_, err := nodes.lite.ChainHead(ctx)
+		req.NoError(err)
+	}
+	tokensUsed := loops * tokensPerLoop
+	expectedEnd := start.Add(time.Duration(float64(tokensUsed) / float64(tokensPerSecond) * float64(time.Second)))
+	allowPad := time.Duration(float64(tokensPerLoop*2) / float64(tokensPerSecond) * float64(time.Second)) // add padding to account for slow test runs
+	t.Logf("expected end: %s, now: %s, allowPad: %s, actual delta: %s", expectedEnd, time.Now(), allowPad, time.Since(expectedEnd))
```

Comment:
I witnessed a flaky on this test, so I loosened it up a bit; I extended the loops to 25 to smooth out the bumps and doubled the padding. Mostly I've witnessed the expected end be within about 50ms of actual on my machine.

Reply:
https://github.com/filecoin-project/lotus/actions/runs/10300409613/job/28509756058?pr=12327

```
    gateway_test.go:398: expected end: 2024-08-08 10:32:18.112474055 +0000 UTC m=+21.555100385, now: 2024-08-08 10:32:18.167164355 +0000 UTC m=+21.609790693, allowPad: 800ms, actual delta: 54.690384ms
```